File: packages/admin-console/src/api/branches.ts
import { api } from './client'

const DEFAULT_TENANT = 'default'

export interface Branch {
  name: string
  head: string  // HLC format: "timestamp-counter"
  created_at: string
  created_by: string
  protected: boolean
  parent_revision?: string
  upstream_branch?: string  // Branch to compare against for divergence
}

export interface BranchDivergence {
  ahead: number  // Number of commits ahead
  behind: number  // Number of commits behind
  common_ancestor: string  // HLC format
}

export type MergeStrategy = 'FastForward' | 'ThreeWay'

export type ConflictType =
  | 'BothModified'
  | 'DeletedBySourceModifiedByTarget'
  | 'ModifiedBySourceDeletedByTarget'
  | 'BothAdded'

export interface MergeConflict {
  node_id: string
  path: string
  conflict_type: ConflictType
  base_properties: any | null
  target_properties: any | null
  source_properties: any | null
  /** Translation locale if this is a translation conflict (undefined = base node conflict) */
  translation_locale?: string
}

export interface MergeResult {
  success: boolean
  revision: string | null  // HLC format
  conflicts: MergeConflict[]
  fast_forward: boolean
  nodes_changed: number
}

export interface MergeBranchRequest {
  source_branch: string
  strategy: MergeStrategy
  message: string
  actor: string
}

export interface ConflictResolution {
  node_id: string
  resolution_type: 'keep-ours' | 'keep-theirs' | 'manual'
  resolved_properties: any
  /** Translation locale being resolved (undefined = base node conflict) */
  translation_locale?: string
}

export interface ResolveMergeRequest {
  source_branch: string
  resolutions: ConflictResolution[]
  message: string
  actor: string
}

export interface Tag {
  name: string
  revision: string  // HLC format: "timestamp-counter"
  created_at: string
  created_by: string
  message?: string
  protected: boolean
}

export interface HeadResponse {
  revision: string  // HLC format
}

export interface CreateBranchRequest {
  name: string
  from_revision?: string  // HLC format
  upstream_branch?: string  // Branch to compare against for divergence
  created_by?: string
  protected?: boolean
  include_revision_history?: boolean  // Default true - copy revision history from source branch
}

export interface CreateTagRequest {
  name: string
  revision: string  // HLC format
  created_by?: string
  message?: string
  protected?: boolean
}

export interface UpdateBranchHeadRequest {
  revision: string  // HLC format
}

export const branchesApi = {
  /**
   * List all branches in a repository
   * GET /api/management/repositories/{tenant}/{repo}/branches
   */
  list: (repoId: string) =>
    api.get<Branch[]>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/branches`
    ),

  /**
   * Get a specific branch
   * GET /api/management/repositories/{tenant}/{repo}/branches/{name}
   */
  get: (repoId: string, name: string) =>
    api.get<Branch>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/branches/${name}`
    ),

  /**
   * Create a new branch
   * POST /api/management/repositories/{tenant}/{repo}/branches
   */
  create: (repoId: string, data: CreateBranchRequest) =>
    api.post<Branch>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/branches`,
      data
    ),

  /**
   * Delete a branch
   * DELETE /api/management/repositories/{tenant}/{repo}/branches/{name}
   */
  delete: (repoId: string, name: string) =>
    api.delete<void>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/branches/${name}`
    ),

  /**
   * Get branch HEAD revision (structured response)
   * GET /api/management/repositories/{tenant}/{repo}/branches/{name}/head
   */
  getHead: (repoId: string, name: string) =>
    api.get<HeadResponse>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/branches/${name}/head`
    ),
  
  /**
   * Get branch HEAD revision (convenience method)
   */
  getHeadRevision: async (repoId: string, name: string): Promise<string> => {
    const response = await branchesApi.getHead(repoId, name)
    return response.revision
  },

  /**
   * Update branch HEAD pointer
   * PUT /api/management/repositories/{tenant}/{repo}/branches/{name}/head
   */
  updateHead: (repoId: string, name: string, revision: number) =>
    api.put<void>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/branches/${name}/head`,
      { revision }
    ),

  /**
   * Compare two branches to calculate divergence (commits ahead/behind)
   * GET /api/management/repositories/{tenant}/{repo}/branches/{branch}/compare/{baseBranch}
   */
  compare: (repoId: string, branch: string, baseBranch: string) =>
    api.get<BranchDivergence>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/branches/${branch}/compare/${baseBranch}`
    ),

  /**
   * Merge a source branch into a target branch
   * POST /api/management/repositories/{tenant}/{repo}/branches/{targetBranch}/merge
   */
  merge: (repoId: string, targetBranch: string, data: MergeBranchRequest) =>
    api.post<MergeResult>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/branches/${targetBranch}/merge`,
      data
    ),

  /**
   * Resolve merge conflicts and complete the merge
   * POST /api/management/repositories/{tenant}/{repo}/branches/{targetBranch}/resolve-merge
   */
  resolveMerge: (repoId: string, targetBranch: string, data: ResolveMergeRequest) =>
    api.post<MergeResult>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/branches/${targetBranch}/resolve-merge`,
      data
    ),

  /**
   * Set the upstream branch for divergence comparison
   * PATCH /api/management/repositories/{tenant}/{repo}/branches/{name}/upstream
   */
  setUpstream: (repoId: string, branchName: string, upstreamBranch: string | null) =>
    api.patch<Branch>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/branches/${branchName}/upstream`,
      { upstream_branch: upstreamBranch }
    ),
}

export const tagsApi = {
  /**
   * List all tags in a repository
   * GET /api/management/repositories/{tenant}/{repo}/tags
   */
  list: (repoId: string) =>
    api.get<Tag[]>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/tags`
    ),

  /**
   * Get a specific tag
   * GET /api/management/repositories/{tenant}/{repo}/tags/{name}
   */
  get: (repoId: string, name: string) =>
    api.get<Tag>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/tags/${name}`
    ),

  /**
   * Create a new tag
   * POST /api/management/repositories/{tenant}/{repo}/tags
   */
  create: (repoId: string, data: CreateTagRequest) =>
    api.post<Tag>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/tags`,
      data
    ),

  /**
   * Delete a tag
   * DELETE /api/management/repositories/{tenant}/{repo}/tags/{name}
   */
  delete: (repoId: string, name: string) =>
    api.delete<void>(
      `/api/management/repositories/${DEFAULT_TENANT}/${repoId}/tags/${name}`
    ),
}

File: crates/raisin-transport-http/src/handlers/branches.rs
// SPDX-License-Identifier: BSL-1.1

//! HTTP handlers for Branch management
//!
//! Provides REST API endpoints for:
//! - Creating branches
//! - Listing branches
//! - Getting branch information
//! - Deleting branches
//! - Getting/updating branch HEAD

use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};

use raisin_context::{Branch, BranchDivergence, MergeResult, MergeStrategy};
use raisin_hlc::HLC;
use raisin_storage::{BranchRepository, Storage};

use crate::{error::ApiError, state::AppState};

/// Request body for creating a branch
#[derive(serde::Deserialize)]
pub struct CreateBranchRequest {
    /// Name for the new branch
    pub name: String,
    /// Optional revision to branch from (None = create from scratch)
    pub from_revision: Option<HLC>,
    /// Optional upstream branch for divergence comparison
    pub upstream_branch: Option<String>,
    /// Actor creating the branch
    pub created_by: Option<String>,
    /// Whether the branch is protected from deletion
    #[serde(default)]
    pub protected: bool,
    /// Whether to include revision history from the source branch (default: true)
    /// When enabled, revision metadata is copied in a background job
    #[serde(default = "default_true")]
    pub include_revision_history: bool,
}

fn default_true() -> bool {
    true
}

/// Request body for updating branch HEAD
#[derive(serde::Deserialize)]
pub struct UpdateBranchHeadRequest {
    /// New HEAD revision number
    pub revision: HLC,
}

/// Request body for merging branches
#[derive(serde::Deserialize)]
pub struct MergeBranchRequest {
    /// Name of the source branch to merge from
    pub source_branch: String,
    /// Merge strategy to use
    pub strategy: MergeStrategy,
    /// Commit message for the merge
    pub message: String,
    /// Actor performing the merge
    pub actor: String,
}

/// Request body for resolving merge conflicts
#[derive(serde::Deserialize)]
pub struct ResolveMergeRequest {
    /// Name of the source branch being merged from
    pub source_branch: String,
    /// User's resolutions for each conflict
    pub resolutions: Vec<raisin_context::ConflictResolution>,
    /// Commit message for the merge
    pub message: String,
    /// Actor performing the merge
    pub actor: String,
}

/// Response for getting branch HEAD
#[derive(serde::Serialize)]
pub struct HeadResponse {
    /// Current HEAD revision number
    pub revision: HLC,
}

/// Create a new branch
///
/// POST /api/management/repositories/{tenant_id}/{repo_id}/branches
#[axum::debug_handler]
pub async fn create_branch(
    State(state): State<AppState>,
    Path((tenant_id, repo_id)): Path<(String, String)>,
    Json(req): Json<CreateBranchRequest>,
) -> Result<(StatusCode, Json<Branch>), ApiError> {
    let created_by = req.created_by.as_deref().unwrap_or("system");

    let branch = state
        .storage()
        .branches()
        .create_branch(
            &tenant_id,
            &repo_id,
            &req.name,
            created_by,
            req.from_revision,
            req.upstream_branch,
            req.protected,
            req.include_revision_history,
        )
        .await?;

    Ok((StatusCode::CREATED, Json(branch)))
}

/// List all branches in a repository
///
/// GET /api/management/repositories/{tenant_id}/{repo_id}/branches
pub async fn list_branches(
    State(state): State<AppState>,
    Path((tenant_id, repo_id)): Path<(String, String)>,
) -> Result<Json<Vec<Branch>>, ApiError> {
    let branches = state
        .storage()
        .branches()
        .list_branches(&tenant_id, &repo_id)
        .await?;

    Ok(Json(branches))
}

/// Get a specific branch
///
/// GET /api/management/repositories/{tenant_id}/{repo_id}/branches/{name}
pub async fn get_branch(
    State(state): State<AppState>,
    Path((tenant_id, repo_id, name)): Path<(String, String, String)>,
) -> Result<Json<Branch>, ApiError> {
    let branch = state
        .storage()
        .branches()
        .get_branch(&tenant_id, &repo_id, &name)
        .await?;

    match branch {
        Some(b) => Ok(Json(b)),
        None => Err(ApiError::branch_not_found(&name)),
    }
}

/// Delete a branch
///
/// DELETE /api/management/repositories/{tenant_id}/{repo_id}/branches/{name}
pub async fn delete_branch(
    State(state): State<AppState>,
    Path((tenant_id, repo_id, name)): Path<(String, String, String)>,
) -> Result<StatusCode, ApiError> {
    state
        .storage()
        .branches()
        .delete_branch(&tenant_id, &repo_id, &name)
        .await?;

    Ok(StatusCode::NO_CONTENT)
}

/// Get the HEAD revision of a branch
///
/// GET /api/management/repositories/{tenant_id}/{repo_id}/branches/{name}/head
pub async fn get_branch_head(
    State(state): State<AppState>,
    Path((tenant_id, repo_id, name)): Path<(String, String, String)>,
) -> Result<Json<HeadResponse>, ApiError> {
    let branch = state
        .storage()
        .branches()
        .get_branch(&tenant_id, &repo_id, &name)
        .await?;

    match branch {
        Some(b) => Ok(Json(HeadResponse { revision: b.head })),
        None => Err(ApiError::branch_not_found(&name)),
    }
}

/// Update the HEAD revision of a branch
///
/// PUT /api/management/repositories/{tenant_id}/{repo_id}/branches/{name}/head
pub async fn update_branch_head(
    State(state): State<AppState>,
    Path((tenant_id, repo_id, name)): Path<(String, String, String)>,
    Json(req): Json<UpdateBranchHeadRequest>,
) -> Result<StatusCode, ApiError> {
    state
        .storage()
        .branches()
        .update_head(&tenant_id, &repo_id, &name, req.revision)
        .await?;

    Ok(StatusCode::NO_CONTENT)
}

/// Compare two branches to calculate divergence (commits ahead/behind)
///
/// GET /api/management/repositories/{tenant_id}/{repo_id}/branches/{branch}/compare/{base_branch}
///
/// Returns the number of commits the branch is ahead and behind the base branch,
/// similar to GitHub's branch comparison feature.
#[cfg(feature = "storage-rocksdb")]
pub async fn compare_branches(
    State(state): State<AppState>,
    Path((tenant_id, repo_id, branch, base_branch)): Path<(String, String, String, String)>,
) -> Result<Json<BranchDivergence>, ApiError> {
    // Access the concrete RocksDB storage to call calculate_divergence
    let rocksdb_storage = state
        .rocksdb_storage
        .as_ref()
        .ok_or_else(|| ApiError::internal("RocksDB storage not available"))?;

    let divergence = rocksdb_storage
        .branches_impl()
        .calculate_divergence(&tenant_id, &repo_id, &branch, &base_branch)
        .await?;

    Ok(Json(divergence))
}

/// Merge a source branch into a target branch
///
/// POST /api/management/repositories/{tenant_id}/{repo_id}/branches/{target_branch}/merge
///
/// Performs a Git-like merge operation between two branches. Supports both fast-forward
/// and three-way merge strategies. Returns conflict information if the merge cannot be
/// completed automatically.
///
/// # Request Body
/// - `source_branch`: Name of the branch to merge from
/// - `strategy`: Merge strategy (FastForward or ThreeWay)
/// - `message`: Commit message for the merge
/// - `actor`: User performing the merge
///
/// # Response
/// Returns `MergeResult` containing:
/// - `success`: Whether the merge completed successfully
/// - `revision`: Revision number of the merge commit (if successful)
/// - `conflicts`: List of conflicts (if any)
/// - `fast_forward`: Whether this was a fast-forward merge
/// - `nodes_changed`: Number of nodes affected by the merge
#[cfg(feature = "storage-rocksdb")]
pub async fn merge_branches(
    State(state): State<AppState>,
    Path((tenant_id, repo_id, target_branch)): Path<(String, String, String)>,
    Json(req): Json<MergeBranchRequest>,
) -> Result<Json<MergeResult>, ApiError> {
    // Access the concrete RocksDB storage to call merge_branches
    let rocksdb_storage = state
        .rocksdb_storage
        .as_ref()
        .ok_or_else(|| ApiError::internal("RocksDB storage not available"))?;

    let result = rocksdb_storage
        .branches_impl()
        .merge_branches(
            &tenant_id,
            &repo_id,
            &target_branch,
            &req.source_branch,
            req.strategy,
            &req.message,
            &req.actor,
        )
        .await?;

    Ok(Json(result))
}

/// Resolve merge conflicts and complete the merge
///
/// POST /api/management/repositories/{tenant_id}/{repo_id}/branches/{target_branch}/resolve-merge
///
/// This endpoint is called after `merge_branches` has detected conflicts. The client
/// provides resolutions for all conflicts, and this endpoint creates the merge commit
/// with the resolved state.
///
/// # Request Body
/// - `source_branch`: Name of the branch being merged from
/// - `resolutions`: Array of conflict resolutions, each containing:
///   - `node_id`: ID of the conflicted node
///   - `resolution_type`: Type of resolution (KeepOurs, KeepTheirs, Manual)
///   - `resolved_properties`: The final properties for the node
/// - `message`: Commit message for the merge
/// - `actor`: User performing the merge resolution
///
/// # Response
/// Returns `MergeResult` containing:
/// - `success`: true (since conflicts were resolved)
/// - `revision`: Revision number of the merge commit
/// - `conflicts`: Empty array (all conflicts resolved)
/// - `fast_forward`: false (this is always a three-way merge)
/// - `nodes_changed`: Number of nodes affected by the merge
///
/// # Errors
/// - 404 if either branch doesn't exist
/// - 500 if the merge commit creation fails
#[cfg(feature = "storage-rocksdb")]
pub async fn resolve_merge_conflicts(
    State(state): State<AppState>,
    Path((tenant_id, repo_id, target_branch)): Path<(String, String, String)>,
    Json(req): Json<ResolveMergeRequest>,
) -> Result<Json<MergeResult>, ApiError> {
    // Access the concrete RocksDB storage
    let rocksdb_storage = state
        .rocksdb_storage
        .as_ref()
        .ok_or_else(|| ApiError::internal("RocksDB storage not available"))?;

    let result = rocksdb_storage
        .branches_impl()
        .resolve_merge_with_resolutions(
            &tenant_id,
            &repo_id,
            &target_branch,
            &req.source_branch,
            req.resolutions,
            &req.message,
            &req.actor,
        )
        .await?;

    Ok(Json(result))
}

/// Request body for setting upstream branch
#[derive(serde::Deserialize)]
pub struct SetUpstreamRequest {
    /// The upstream branch name, or null to clear
    pub upstream_branch: Option<String>,
}

/// Set the upstream branch for divergence comparison
///
/// PATCH /api/management/repositories/{tenant_id}/{repo_id}/branches/{name}/upstream
///
/// Sets which branch this branch should be compared against for divergence calculation.
/// When upstream is None, the default branch (usually "main") is used for comparison.
#[cfg(feature = "storage-rocksdb")]
pub async fn set_upstream_branch(
    State(state): State<AppState>,
    Path((tenant_id, repo_id, name)): Path<(String, String, String)>,
    Json(req): Json<SetUpstreamRequest>,
) -> Result<Json<Branch>, ApiError> {
    let rocksdb_storage = state
        .rocksdb_storage
        .as_ref()
        .ok_or_else(|| ApiError::internal("RocksDB storage not available"))?;

    let branch = rocksdb_storage
        .branches_impl()
        .set_upstream_branch(&tenant_id, &repo_id, &name, req.upstream_branch.as_deref())
        .await?;

    Ok(Json(branch))
}
